WebM player should detect resolution size change.

jyavenard opened the pull request:
#### 5e2d4bff9728c23b33dd01b4e84ba78355858b98
<pre>
WebM player should detect resolution size change.
<a href="https://bugs.webkit.org/show_bug.cgi?id=298653">https://bugs.webkit.org/show_bug.cgi?id=298653</a>
<a href="https://rdar.apple.com/160280571">rdar://160280571</a>

Reviewed by Jer Noble.

We adopt the same heuristic as the one in use in the MediaPlayerPrivateMediaSourceAVFObjC
and detect when resolution occurs and provides a callback at the time the change happens.

Test added.

* LayoutTests/media/content/300then400.webm: Added.
* LayoutTests/media/media-vp9-reschange-webm-expected.txt: Added.
* LayoutTests/media/media-vp9-reschange-webm.html: Added.
* Source/WebCore/platform/MediaSample.h:
* Source/WebCore/platform/graphics/AudioVideoRenderer.h:
(WebCore::VideoInterface::notifySizeChanged):
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::convertEnumerationToString):
* Source/WebCore/platform/graphics/MediaPlayerEnums.h: Move enum to MediaPlayerEnums to reduce number of header to include.
(WTF::LogArgument&lt;WebCore::TrackInfoTrackType&gt;::toString):
* Source/WebCore/platform/graphics/AudioVideoRenderer.h:
(WebCore::VideoInterface::notifySizeChanged):
* Source/WebCore/platform/graphics/PlatformTimeRanges.cpp:
(WebCore::PlatformTimeRanges::containWithEpsilon const): Add convenience method
(WebCore::PlatformTimeRanges::findWithEpsilon const): Make method const
(WebCore::PlatformTimeRanges::findWithEpsilon): Deleted.
* Source/WebCore/platform/graphics/PlatformTimeRanges.h:
* Source/WebCore/platform/graphics/avfoundation/AudioVideoRendererAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/AudioVideoRendererAVFObjC.mm:
(WebCore::AudioVideoRendererAVFObjC::enqueueSample):
(WebCore::AudioVideoRendererAVFObjC::notifySizeChanged):
(WebCore::AudioVideoRendererAVFObjC::sizeWillChangeAtTime):
(WebCore::AudioVideoRendererAVFObjC::flushPendingSizeChanges):
(WebCore::AudioVideoRendererAVFObjC::flushVideo):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::MediaPlayerPrivateWebM):
(WebCore::MediaPlayerPrivateWebM::waitForTimeBuffered): The test identified an issue when
the first sample doesn&apos;t start with a time of 0.
As such, we couldn&apos;t seek to 0 as we would wait for a future frame that would never came.
We use a fuzzy search that allows 10ms gap, similar to what MSE player is doing.
(WebCore::MediaPlayerPrivateWebM::enqueueSample):
(WebCore::MediaPlayerPrivateWebM::appendCompleted):
(WebCore::MediaPlayerPrivateWebM::trackDidChangeEnabled):

Canonical link: <a href="https://commits.webkit.org/299831@main">https://commits.webkit.org/299831@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/23da4d9a844bb90bb3466a1c3346e610119b9bfe

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/120390 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/40084 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/30736 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/126761 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/72465 "Built successfully") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/4ba3969f-d246-459a-b691-16f28e852172) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/122266 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/40781 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/48661 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/91432 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/60727 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/f209f887-7097-4de8-862d-9cc17fae5c17) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/123342 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/32569 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107925 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71984 "Passed tests") | | [❌ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/230e78b6-6bb0-467a-83bd-3a94ba3933bd) 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/119747 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31603 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/26030 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/70377 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/102050 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/26214 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/129646 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/47311 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/35902 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/100051 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/47677 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/104108 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99893 "Passed tests") | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/25347 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45342 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23373 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/43957 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/47173 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/52878 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/46641 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49988 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/48328 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

jernoble commented on Source/WebCore/platform/graphics/avfoundation/AudioVideoRendererAVFObjC.mm:
```diff
@@ -157,11 +157,37 @@ @interface AVSampleBufferDisplayLayer (Staging_100128644)
+        }
+
+        if (m_sizeChangedCallback) {
+            FloatSize formatSize = FloatSize(PAL::CMVideoFormatDescriptionGetPresentationDimensions(formatDescription, true, true));
```
Nit: now that we have it, we should probably use `FormatDescriptionUtilities` here.

jyavenard committed:
WebM player should detect resolution size change.
https://bugs.webkit.org/show_bug.cgi?id=298653
rdar://160280571

Reviewed by Jer Noble.

We adopt the same heuristic as the one in use in the MediaPlayerPrivateMediaSourceAVFObjC
and detect when resolution occurs and provides a callback at the time the change happens.

Test added.

* LayoutTests/media/content/300then400.webm: Added.
* LayoutTests/media/media-vp9-reschange-webm-expected.txt: Added.
* LayoutTests/media/media-vp9-reschange-webm.html: Added.
* Source/WebCore/platform/MediaSample.h:
* Source/WebCore/platform/graphics/AudioVideoRenderer.h:
(WebCore::VideoInterface::notifySizeChanged):
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::convertEnumerationToString):
* Source/WebCore/platform/graphics/MediaPlayerEnums.h: Move enum to MediaPlayerEnums to reduce number of header to include.
(WTF::LogArgument<WebCore::TrackInfoTrackType>::toString):
* Source/WebCore/platform/graphics/AudioVideoRenderer.h:
(WebCore::VideoInterface::notifySizeChanged):
* Source/WebCore/platform/graphics/PlatformTimeRanges.cpp:
(WebCore::PlatformTimeRanges::containWithEpsilon const): Add convenience method
(WebCore::PlatformTimeRanges::findWithEpsilon const): Make method const
(WebCore::PlatformTimeRanges::findWithEpsilon): Deleted.
* Source/WebCore/platform/graphics/PlatformTimeRanges.h:
* Source/WebCore/platform/graphics/avfoundation/AudioVideoRendererAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/AudioVideoRendererAVFObjC.mm:
(WebCore::AudioVideoRendererAVFObjC::enqueueSample):
(WebCore::AudioVideoRendererAVFObjC::notifySizeChanged):
(WebCore::AudioVideoRendererAVFObjC::sizeWillChangeAtTime):
(WebCore::AudioVideoRendererAVFObjC::flushPendingSizeChanges):
(WebCore::AudioVideoRendererAVFObjC::flushVideo):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::MediaPlayerPrivateWebM):
(WebCore::MediaPlayerPrivateWebM::waitForTimeBuffered): The test identified an issue when
the first sample doesn't start with a time of 0.
As such, we couldn't seek to 0 as we would wait for a future frame that would never came.
We use a fuzzy search that allows 10ms gap, similar to what MSE player is doing.
(WebCore::MediaPlayerPrivateWebM::enqueueSample):
(WebCore::MediaPlayerPrivateWebM::appendCompleted):
(WebCore::MediaPlayerPrivateWebM::trackDidChangeEnabled):

Canonical link: https://commits.webkit.org/299831@main
